src/Storages/ObjectStorage/StorageObjectStorage.cpp

```diff
@@ -502,6 +586,34 @@ SinkToStoragePtr StorageObjectStorage::write(
         local_context);
 }
 
+bool StorageObjectStorage::optimize(
+    const ASTPtr & /*query*/,
+    [[maybe_unused]] const StorageMetadataPtr & metadata_snapshot,
+    const ASTPtr & /*partition*/,
+    bool /*final*/,
+    bool /*deduplicate*/,
+    const Names & /* deduplicate_by_columns */,
+    bool /*cleanup*/,
+    [[maybe_unused]] ContextPtr context)
+{
+#if USE_AVRO
+    if (configuration->isDataLakeConfiguration() && configuration->supportsWrites())
```

Comment:
Same here, `iceberg` soon will be not the only data lake supporting writes, so this is incorrect to directly assume iceberg to be everywhere (`allow_experimental_iceberg_compaction` can be enabled while this storage does not have to be iceberg, so we can end up executing iceberg code for non iceberg storage).